src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java

```diff
@@ -0,0 +1,145 @@
+        .isEmpty()) {
+      throw new RuntimeException("Table name is required");
+    }
+    String tableName = configuration.getString("tableName");
+    JsonObject properties = Json.createObjectBuilder().build();
+    Connection connection = null;
+    ResultSet rs = null;
+    String schemaName = null;
+    Boolean isEmpty = true;
+    Boolean isOracle = (configuration.getString("dbEngine").equals("oracle")) ? true : false;
+    Boolean isMssql = (configuration.getString("dbEngine").equals("mssql")) ? true : false;
+    List<String> primaryKeys = new ArrayList();
+    try {
+      connection = Utils.getConnection(configuration);
+      DatabaseMetaData dbMetaData = connection.getMetaData();
+      if (tableName.contains(".")) {
+        schemaName =
+            (isOracle) ? tableName.split("\\.")[0].toUpperCase() : tableName.split("\\.")[0];
+        tableName =
+            (isOracle) ? tableName.split("\\.")[1].toUpperCase() : tableName.split("\\.")[1];
+      }
+      rs = dbMetaData
+          .getPrimaryKeys(null, ((isOracle && !schemaName.isEmpty()) ? schemaName : null),
+              tableName);
+      while (rs.next()) {
+        primaryKeys.add(rs.getString("COLUMN_NAME"));
+        logger.info("Primary Key: {}", rs.getString("COLUMN_NAME"));
+      }
+      rs = dbMetaData
+          .getColumns(null, ((isOracle && !schemaName.isEmpty()) ? schemaName : null), tableName,
+              "%");
+      while (rs.next()) {
+        if (primaryKeys.contains(rs.getString("COLUMN_NAME"))) {
+          JsonObjectBuilder field = Json.createObjectBuilder();
+          String name = rs.getString("COLUMN_NAME");
+          Boolean isRequired = false;
+          if (isMssql) {
+            String isAutoincrement =
+                (rs.getString("IS_AUTOINCREMENT") != null) ? rs.getString("IS_AUTOINCREMENT") : "";
+            Integer isNullable = (rs.getObject("NULLABLE") != null) ? rs.getInt("NULLABLE") : 1;
+            isRequired = isNullable == 0 && !isAutoincrement.equals("YES");
+          } else {
+            isRequired = true;
+          }
+          field.add("required", isRequired)
+              .add("title", name)
+              .add("type", convertType(rs.getInt("DATA_TYPE")));
+          properties = Json.createObjectBuilder().add(name, field.build()).build();
+          isEmpty = false;
+        }
+      }
+      if (isEmpty) {
+        logger.info("Empty PK list - no primary keys");
+        throw new IllegalStateException("No Primary Keys");
+      }
+
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    } finally {
+      if (rs != null) {
+        try {
+          rs.close();
+        } catch (SQLException e) {
+          logger.error("Failed to close result set", e.toString());
+        }
+      }
+      if (connection != null) {
+        try {
+          connection.close();
+        } catch (SQLException e) {
+          logger.error("Failed to close connection", e.toString());
```

Comment:
logger.error("Failed to close connection {}", e);
where is exception throwing ?

Reply:
no need changes (was discussed)